### PR TITLE
parmetis 4.0.3 (migrated from homebrew-science)

### PR DESCRIPTION
**Pull request by jslee02**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Migrating because of Homebrew/homebrew-science#6331

### CODE REVIEW

**Comment by jslee02** [comment]
This PR is not ready to review. Let me reopen once it's ready.

**Comment by galbramc** [comment]
@jslee02, may I offer to help with migrating this formula? I do research with computational fluid dynamics, and our group relies on parmetis for mesh partitioning.

**Comment by jslee02** [comment]
> @jslee02, may I offer to help with migrating this formula?

Sure, it would be appreciated. I added you as a collaborator to https://github.com/jslee02/homebrew-core. Feel free to update the formula.

**Comment by galbramc** [comment]
@jslee02, what is left to clean up here? Looks like the only automatic complaint are the patches, but those are functional fixes. They were reported to Karypis, and I don't know why they have not been incorporated.

**Comment by jslee02** [comment]
I believe there is nothing left to clean up unless the homebrew maintainers find some. :sweat_smile: Let's see.

**Comment by galbramc** [comment]
I can hopefully fix any problems, unless you beat me to it.

File: Formula/parmetis.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
+class Parmetis < Formula
+  desc "MPI library for graph/mesh partitioning and fill-reducing orderings"
+  homepage "http://glaros.dtc.umn.edu/gkhome/metis/parmetis/overview"
+  url "http://glaros.dtc.umn.edu/gkhome/fetch/sw/parmetis/parmetis-4.0.3.tar.gz"
+  sha256 "f2d9a231b7cf97f1fee6e8c9663113ebf6c240d407d3c118c55b3633d6be6e5f"
+
+  # METIS 5.* is required. It comes bundled with ParMETIS.
+  # We prefer to brew it ourselves.
```
Comment (DomT4) [inline Formula/parmetis.rb]: You can simplify these two lines into something like `# Prefer our metis over the bundled one`.

File: Formula/parmetis.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  depends_on "metis"
+
+  depends_on "cmake" => :build
+  depends_on "open-mpi"
+
+  # Do not build the METIS 5.* that ships with ParMETIS.
```
Comment (DomT4) [inline Formula/parmetis.rb]: Should ask upstream for a flag that does this, ideally.

File: Formula/parmetis.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  patch :DATA
+
+  # Bug fixes from PETSc developers. Mirrored because the SHA-256s get
+  # invalidated every time Bitbucket updates the Git version they use.
+  patch do
+    # From: https://bitbucket.org/petsc/pkg-parmetis/commits/82409d68aa1d6cbc70740d0f35024aae17f7d5cb/raw/
```
Comment (DomT4) [inline Formula/parmetis.rb]: This was committed in 2012, how hasn't it made it into a stable release yet? 😅.

File: Formula/parmetis.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  patch do
+    # From: https://bitbucket.org/petsc/pkg-parmetis/commits/1c1a9fd0f408dc4d42c57f5c3ee6ace411eb222b/raw/
+    url "https://raw.githubusercontent.com/Homebrew/patches/f104fbb1e09402798cbbc06d2f695d85398c0c89/parmetis/commit-1c1a9fd0.patch"
+    sha256 "baec5e1fa6bb4f6c59e3ede564485e0ad743f58c9875fd65cb715b5c14a491b5"
+  end
+
+  def install
+    ENV["LDFLAGS"] = "-L#{Formula["metis"].lib} -lmetis -lm"
```
Comment (DomT4) [inline Formula/parmetis.rb]: `.opt_lib` rather than `.lib`

File: Formula/parmetis.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    system "make", "config", "prefix=#{prefix}", "shared=1"
+    system "make", "install"
+    pkgshare.install "Graphs" # Sample data for test
+  end
+
+  test do
+    system "mpirun", "#{bin}/ptest", "#{pkgshare}/Graphs/rotor.graph"
+    ohai "Test results are in ~/Library/Logs/Homebrew/parmetis."
```
Comment (DomT4) [inline Formula/parmetis.rb]: Don't `ohai` in tests. The path used here also cannot be taken for granted; it's a user-controlled variable.

**Comment by galbramc** [comment]
Hopefully I got all the fixes correct. Unfortunately, given that they still have not incorporated the patches, I don't have much hope of them including a cake flag to not build metis.

File: Formula/parmetis.rb (hunk 3 above)
Comment (galbramc) [reply Formula/parmetis.rb]: My best guess is the student graduated, and the project didn't progress much after that.

**Comment by ilovezfs** [comment]
New formulae must build without any patches. Thanks for the PR nonetheless @jslee02

**Comment by galbramc** [comment]
@ilovezfs A large number of scientific software uses metis/parmetis. Removing it from homebrew will prevent other scientific formulas from being built. Is there any compromise that can be reached here?

I can remove the patches if that is the sticking point. They only fix functionality in the code, and clearly many people use parmetis even with these bugs.

**Comment by ilovezfs** [comment]
>A large number of scientific software uses metis/parmetis. Removing it from homebrew will prevent other scientific formulas from being built. Is there any compromise that can be reached here?

It's not in Homebrew.

>I can remove the patches if that is the sticking point. They only fix functionality in the code, and clearly many people use parmetis even with these bugs.

Why have they not been upstreamed if they're fixing things? Is upstream not doing new releases?

**Comment by galbramc** [comment]
They were in homebrew/science, and a large number of scientific research simply uses parmetis as one of their dependencies for research codes that are only used within academic and government labs. It looks like parmetis has not been touched in many years (I'm guessing the student working on the project graduated.) The patches were reported back on 2012 (http://glaros.dtc.umn.edu/gkhome/node/837). Even without the patches, I know a lot of both academic and government researches that use parmetis for their projects.

**Comment by ilovezfs** [comment]
If there is no active upstream, this is not something we will package. Projects that use it should probably vendor the code they need unless someone is going to step up, fork the project, and become an active upstream, to which we can report bugs, etc.

**Comment by galbramc** [comment]
If I can get a hold of George Karypis and get him to apply the patch can parmetis get added to homebrew?

PETSc, which is a very popular scientific package, depends on parmetis. My own CFD research code also uses parmetis. Having parmetis as part of homebrew really helps our community a lot.

**Comment by ilovezfs** [comment]
@galbramc I'm not sure how we can even comply with this license:

>Copyright & License Notice
>--
>ParMETIS is copyrighted by the Regents of the University of Minnesota.  It can be freely used for educational and research purposes by  non-profit institutions and US government agencies only. Other  organizations are allowed to use ParMETIS only for evaluation purposes,  and any further uses will require prior approval. The software may not  be sold or redistributed without prior approval. One may make copies of  the software for their use provided that the copies, are not sold or  distributed, are used under the same terms and conditions.   As unestablished research software, this code is provided on an "as is''  basis without warranty of any kind, either expressed or implied. The  downloading, or executing any part of this software constitutes an  implicit agreement to these terms. These terms and conditions are  subject to change at any time without prior notice.

**Comment by galbramc** [comment]
Oh, is it the line "The software may not be sold or redistributed without prior approval." that is problematic?

**Comment by ilovezfs** [comment]
Yup.

**Comment by galbramc** [comment]
So is Ubuntu technically should not distribute it either?

**Comment by ilovezfs** [comment]
@galbramc It's not for me to say what others should or shouldn't do in this case :)

**Comment by galbramc** [comment]
Fair enough, so I can see distributing bottles might be problematic. But could a formula that compiles from source be acceptable? (Can you tell I would really like to have this :P)

**Comment by ilovezfs** [comment]
We don't allow `bottle :disable` on new formulae.

**Comment by galbramc** [comment]
Dang... so there is nothing I can do to get this into homebrew? (Maybe I can turn it into an old formula...)

**Comment by ilovezfs** [comment]
@galbramc you can put it in [your own tap](https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md#how-to-create-and-maintain-a-tap) but nothing in https://github.com/Homebrew/homebrew-core will be able to depend on it.

Ideally, someone should obtain the approval for Homebrew to distribute it, and get those patches upstreamed into a new tarball release. Call me crazy.

**Comment by galbramc** [comment]
Oh you are not crazy, but being in academia I know how these things works. Government and industry doesn't give out grands to funding code maintenance, only developing new things. I don't agree with it, but life isn't always so rational.

I can make my own tap and that should work for our group. We use parmetis internally, as well as PETSc. PETSc can be compiled without parmetis, but you don't get the partitioning functionality of PETSc without parmetis. I parmetis directly, so it works for me. I was thinking about putting together a formula for zoltan as well (http://www.cs.sandia.gov/zoltan/Zoltan.html), but it also depends on parmetis.

**Comment by ilovezfs** [comment]
>If I can get a hold of George Karypis and get him to apply the patch can parmetis get added to homebrew?

In any case, if you can get the patches upstreamed and get him to release a new tarball, we can revisit this.

**Comment by galbramc** [comment]
Ok. I'll see if I can reach Dr. Karypis and see what we can do. I don't know him but it's worth a try.

**Comment by fxcoudert** [comment]
Academic software has crazy licenses :(
And goes unmaintained when people switch interest :(

I suggest a private tap, indeed. Many people can use it!

**Pull request closed without merging**